### PR TITLE
Adding V2 of bicep files for sample Open AI Infra

### PR DESCRIPTION
**Pull request by snorris31**
## Purpose
Sets up Open AI in a secured environment using private endpoints and VNET (excluding observability)
image

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

```
git clone https://github.com/Azure-Samples/openai-enterprise-iac.git
cd openai-enterprise-iac
git checkout sanorris/openaibicep
```

## What to Check
- Use README to create resource group and deploy

### CODE REVIEW

**Comment by masa-fleet** [comment]
Just curious, but is bicep module file naming convention snake case?

File: infra/main.bicep [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+// Licensed under the MIT License.
+
+@description('The project name or prefix for all resources.')
+param name string = 'openaiproject'
+
+@description('The location where the resource is created.')
+param location string = resourceGroup().location
+
+@description('SKU value')
+param sku string = 'S0'
```
Comment (masa-fleet) [inline infra/main.bicep]: Can we use more specific name such as `openAiSku`?

File: infra/main.bicep [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+@description('VNET Name')
+param vnetName string = '${resourceGroup().name}-vnet'
+
+@description('Subnet Name')
+param subnetName string = '${resourceGroup().name}-subnet'
+
+@description('Subnet Name')
+param pepSubnetName string = '${resourceGroup().name}-pep-subnet'
```
Comment (masa-fleet) [inline infra/main.bicep]: Can we add more description to understand what is `pepSubnet`? Audiences may not understand what is `pep`.

File: infra/main.bicep [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+@description('The project name or prefix for all resources.')
+param name string = 'openaiproject'
+
+@description('The location where the resource is created.')
+param location string = resourceGroup().location
+
+@description('SKU value')
+param sku string = 'S0'
+
```
Comment (masa-fleet) [inline infra/main.bicep]: Just curious, but why we don't specify appService SKU here even though we specify OpenAI SKU here?

File: README.md [hunk 4]
```diff
@@ -1,52 +1,24 @@
 # Project Name
 
-(short, 1-3 sentenced, description of the project)
+Sets up sample Open AI infrastructure with security considered
```
Comment (kenakamu) [inline README.md]:
```suggestion
Sets up a sample Azure Open AI infrastructure with security considered
```

File: README.md [hunk 5]
```diff
@@ -1,52 +1,24 @@
...
 
 ## Features
 
 This project framework provides the following features:
 
-* Feature 1
-* Feature 2
-* ...
+- Sets up app service, Open AI service, cognitive search service, private endpoint, and VNET
```
Comment (kenakamu) [inline README.md]:
```suggestion
- Sets up app service, Azure Open AI service, cognitive search service, private endpoint, and VNet
```

File: README.md [hunk 6]
```diff
@@ -1,52 +1,24 @@
...
-- ...
-
-### Installation
-
-(ideally very short)
-
-- npm install [package name]
-- mvn install
-- ...
+- Add bicep extension to VSCode (if developing)
```
Comment (kenakamu) [inline README.md]:
```suggestion
- Add bicep extension to Visual Studio Code (if developing)
```

File: README.md [hunk 7]
```diff
@@ -1,52 +1,24 @@
...
 
 ### Quickstart
-(Add steps to get up and running quickly)
 
 1. git clone [repository clone url]
```
Comment (kenakamu) [inline README.md]: update the url

File: infra/app_service.bicep [hunk 8]
```diff
@@ -0,0 +1,36 @@
+param appName string
+param appServicePlanName string
+param location string
+param vnetName string
+param subnetName string
+
+resource vnet 'Microsoft.Network/virtualNetworks@2021-08-01' existing = {
```
Comment (kenakamu) [inline infra/app_service.bicep]: Use the latest version if you don't have any specific reason

File: infra/main.bicep [hunk 9]
```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (kenakamu) [inline infra/main.bicep:1]: Add copy right to other files, too.

File: infra/main.bicep [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+// Licensed under the MIT License.
+
+@description('The project name or prefix for all resources.')
+param name string = 'openaiproject'
+
+@description('The location where the resource is created.')
+param location string = resourceGroup().location
+
+@description('SKU value')
```
Comment (kenakamu) [inline infra/main.bicep]: Add more explanation which service we will apply this SKU?

File: infra/main.bicep [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+param sku string = 'S0'
+
+@description('VNET Name')
+param vnetName string = '${resourceGroup().name}-vnet'
+
+@description('Subnet Name')
+param subnetName string = '${resourceGroup().name}-subnet'
+
+@description('Subnet Name')
```
Comment (kenakamu) [inline infra/main.bicep]: Wrong description

File: infra/main.bicep [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+@description('The project name or prefix for all resources.')
+param name string = 'openaiproject'
+
+@description('The location where the resource is created.')
+param location string = resourceGroup().location
+
+@description('SKU value')
+param sku string = 'S0'
+
+@description('VNET Name')
```
Comment (kenakamu) [inline infra/main.bicep]:
```suggestion
@description('VNet Name')
```

File: infra/main.bicep [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+param vnetName string = '${resourceGroup().name}-vnet'
+
+@description('Subnet Name')
+param subnetName string = '${resourceGroup().name}-subnet'
+
+@description('Subnet Name')
+param pepSubnetName string = '${resourceGroup().name}-pep-subnet'
+
+@description('The pricing tier of the search service you want to create (for example, basic or standard).')
```
Comment (kenakamu) [inline infra/main.bicep]: Consider using enum by using '@allowed' keyword

File: infra/main.bicep [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+param cogSearchSku string = 'basic'
+
+@description('The name of the search service')
+param cogServiceName string = '${resourceGroup().name}-cog-search-service-smn'
+
+@description('The open AI name')
```
Comment (kenakamu) [inline infra/main.bicep]:
```suggestion
@description('The name of the Auzre Open AI service')
```

File: infra/main.bicep [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+@description('The name of the app service')
+param appServiceName string = '${resourceGroup().name}-appservice'
+
+@description('The name of the private endpoint')
+param privateEndpointName string = '${resourceGroup().name}-privateendpoint'
+
+@description('The name of the virtual network link')
+param virtualNetworkLinkName string = '${resourceGroup().name}-virtualnetworklink'
+
+module deployVnet './vnet.bicep' = {
```
Comment (kenakamu) [inline infra/main.bicep]: Always use VNet for consistency

Comment (kenakamu) [inline infra/main.bicep]: why add deploy only for VNet?

**Review by kenakamu — CHANGES_REQUESTED** [review]

**Review by masa-fleet — COMMENTED** [review]
Added some comments. Please review and update some contents :)

File: infra/main.bicep [hunk 16]
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+@description('The project name or prefix for all resources.')
+param name string = 'openaiproject'
+
+@description('The location where the resource is created.')
+param location string = resourceGroup().location
+
+@description('Open AI SKU value')
```
Comment (kenakamu) [inline infra/main.bicep]:
```suggestion
@description('Azure Open AI SKU value')
```

File: infra/main.bicep [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+param vnetName string = '${resourceGroup().name}-vnet'
+
+@description('Subnet Name')
+param subnetName string = '${resourceGroup().name}-subnet'
+
+@description('Private endpoint subnet Name')
+param pepSubnetName string = '${resourceGroup().name}-pep-subnet'
+
+@description('The pricing tier of the search service you want to create')
+@allowed([
```
Comment (kenakamu) [inline infra/main.bicep]: cannot create free?

File: infra/main.bicep [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+  'basic'
+  'standard'
+])
+param cogSearchSku string = 'basic'
+
+@description('The name of the search service')
+param cogServiceName string = '${resourceGroup().name}-cog-search-service-smn'
```
Comment (kenakamu) [inline infra/main.bicep]:
```suggestion
param cogSearchName string = '${resourceGroup().name}-cog-search-service-smn'
```

File: infra/main.bicep [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+
+@description('The name of the Open AI service')
```
Comment (kenakamu) [inline infra/main.bicep]:
```suggestion
@description('The name of the Azure Open AI service')
```

File: infra/main.bicep [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+    pepSubnetName: pepSubnetName
+  }
+}
+
+module appService './app_service.bicep' = {
+  name: 'appServices'
+  params: {
+    location: location
+    appName: appServiceName
+    appServicePlanName: 'asp'
```
Comment (kenakamu) [inline infra/main.bicep]: why is this hardcoded?

File: infra/app_service.bicep [hunk 21]
```diff
@@ -0,0 +1,37 @@
+param appName string
```
Comment (kenakamu) [inline infra/app_service.bicep:4]: Just curious why do we need app service?

File: infra/open_ai.bicep [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+param location string
+param openAIName string
+param vnetName string
+param subnetName string
+param resourceGroup string
+param subscriptionId string
+param cogServiceName string
+param cogSearchSku string
+
+resource name_resource 'Microsoft.CognitiveServices/accounts@2022-03-01' = {
```
Comment (kenakamu) [inline infra/open_ai.bicep]: As we separate bicep files per Azure resource, should we separate Cognitive Search and Azure Open AI?

File: infra/open_ai.bicep [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    publicNetworkAccess: 'Enabled'
+    networkAcls: {
+      defaultAction: 'Deny'
+      virtualNetworkRules: json('[{"id": "${subscriptionId}/resourceGroups/${resourceGroup}/providers/Microsoft.Network/virtualNetworks/${vnetName}/subnets/${subnetName}"}]')
+      ipRules: json('[]')
+    }
+  }
+}
+
+resource search 'Microsoft.Search/searchServices@2022-09-01' = {
```
Comment (kenakamu) [inline infra/open_ai.bicep]: Just curious if we need search service?

File: infra/private_endpoint.bicep [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        {
+          name: 'privatelink-search-windows-net'
+          properties: {
+            privateDnsZoneId: dnsZones.id
+          }
+        }
+      ]
+    }
+  }
+
```
Comment (kenakamu) [inline infra/private_endpoint.bicep]: Remove the line

**Comment by snorris31** [comment]
> Just curious, but is bicep module file naming convention snake case?

Most places I see use snake case, but I don't think there is a specific casing requirement!

**Comment by kenakamu** [comment]
Thanks for adding the private endpoint for Open AI. One feedback from me is to re-consider the file structure. For example, private_endpoint.bicep contains Private DNS and private endpoint for Azure Cognitive Search, but private_endpoint_openai.bicep contains private endpoint for Open AI.

Also oepn_ai.bicep contains both Open AI and Cognitive Search. If you want to reduce the file, you can merge private endpoint into one like you are doing so in open_ai.bicep. Otherwise I recommend you to separate files for individual resource.

**Comment by snorris31** [comment]
> Thanks for adding the private endpoint for Open AI. One feedback from me is to re-consider the file structure. For example, private_endpoint.bicep contains Private DNS and private endpoint for Azure Cognitive Search, but private_endpoint_openai.bicep contains private endpoint for Open AI.
> 
> Also oepn_ai.bicep contains both Open AI and Cognitive Search. If you want to reduce the file, you can merge private endpoint into one like you are doing so in open_ai.bicep. Otherwise I recommend you to separate files for individual resource.

sure, I separated open ai and cognitive search

File: infra/main.bicep [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+    location: location
+    cogSearchSku: cogSearchSku
+    cogServiceName: cogSearchName
+  }
+  dependsOn: [
+    vnet
+  ]
+}
+
+module privateEndpoint './private_endpoint_cognitivesearch.bicep' = {
```
Comment (kenakamu) [inline infra/main.bicep]: please change the name to privateEndpointCognitiveSearch

File: infra/main.bicep [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+  name: openAiPrivateEndpointName
+  params: {
+    openAiName: openAIName
+    location: location
+    privateEndpointOpenAiName: openAiPrivateEndpointName
+    openAiSubnetName: openAiPepSubnetName
+    vnetName: vnetName
+  }
+  dependsOn: [
+    openAi
```
Comment (kenakamu) [inline infra/main.bicep]: As DNS is created above, I think we can depends on it as well? Thought bicep is smart enough to figure out the dependency with existing keyword.

**Review by kenakamu — CHANGES_REQUESTED** [review]

File: infra/main.bicep (hunk 26 above)
Comment (snorris31) [reply infra/main.bicep]: done!

File: infra/main.bicep (hunk 25 above)
Comment (snorris31) [reply infra/main.bicep]: done!

**Review by kenakamu — APPROVED** [review]

**Pull request merged**